### DISCUSSION TITLE
FYI: Godot and Supabase CSharp

### DISCUSSION THREAD

**Discussion by wiverson** [supabase-community/supabase-csharp#127] [General]
I had some time today and I decided to test using Godot and supabase-csharp together.

Setup: Godot v4.2.rc1.mono.official [ad72de508]
Rider 2023.2.3 w/ the official [Rider Godot plugin](https://plugins.jetbrains.com/plugin/13882-godot-support).
macOS M2

I just opened up the project and added a few UI elements.

<img width="949" alt="image" src="https://github.com/supabase-community/supabase-csharp/assets/873850/a756b22f-2175-4bef-a936-974ad1ad553d">

I then added the supabase dependency via the Rider NuGet manager.

<img width="294" alt="image" src="https://github.com/supabase-community/supabase-csharp/assets/873850/447007de-ff4a-4c12-a8e1-bf088a19076e">

Below is the C# script I attached to the signup button. I think that it would have been better to attach this script to the base node instead of using ../ to find the other nodes but that's pretty minor / side effect of me starting out with Godot.

...and that's pretty much it. It worked fine.

I'm *extremely* impressed. Getting async/await to work properly and post the data back to the UI in Unity requires a bunch of futzing with the threads (e.g. UniTask, polling in Update to update the UI, context mgmt, etc). This snippet just worked, no async workarounds required. I was very impressed that NuGet just worked. Godot is much, much lighter and smaller than Unity - it launches faster.

I exported the project as a macOS standalone dmg and it was much easier & faster than exporting from Unity. The resulting build was fairly large (~100mb dmg, ~300mb app), but it looks like the bulk of that 300mb was a single 147mb executable plus basically two complete copies of the rest of the app (Intel and Apple Silicon/arm).

I exported the project as iOS (which is currently very experimental in 4.2rc) and got some errors that I haven't worked through yet.

All in all, my two cents: if you are targeting desktop via Godot, everything looks fine. If you are targeting iOS, well, watch the Godot team for news on the Godot iOS C# stuff - I know they are working on this actively right now. In the meantime there is a [GDScript implementation for Supabase](https://github.com/supabase-community/godot-engine.supabase) that might work for iOS and Godot w/o C#, although it looks like this implementation hasn't been updated in a while.

The next thing to do would be to add in the listeners for debug, session and persistence implementations for Godot. To match the [Unity template](https://github.com/wiverson/supabase-unity-template) I suppose a default example of a Supabase manager and nice little place to drop settings.

In any event, I highly recommend checking out Godot - it's a lot of fun. It has a ton of features and nodes that require external packages in Unity. It took me a little while to sort out things like the Remote mode to tweak a running game but all in all it's very clean and modern feeling.

```csharp
using Godot;
using System;
using System.Diagnostics.CodeAnalysis;
using Supabase;
using Supabase.Gotrue;
using static Godot.GD;
using Client = Supabase.Client;

// ReSharper disable once CheckNamespace
public partial class SignUpButton : Button
{
	private Client _supabase;

	private LineEdit _lineEdit;
	private LineEdit _passwordEdit;
	private Label _messageLabel;
	
	// Called when the node enters the scene tree for the first time.
	public override void _Ready()
	{
		_lineEdit = GetNode<LineEdit>("../EmailEdit");
		if (_lineEdit == null)
			Print("Unable to find EmailEdit");
		_passwordEdit = GetNode<LineEdit>("../PassEdit");
		if (_passwordEdit == null)
			Print("Unable to find PasswordEdit");
		_messageLabel = GetNode<Label>("../MessageLabel");
		if (_messageLabel == null)
			Print("Unable to find MessageLabel");
	}

	// Called every frame. 'delta' is the elapsed time since the previous frame.
	public override void _Process(double delta)
	{ }

	[SuppressMessage("ReSharper", "StringLiteralTypo")]
	private async void _on_pressed()
	{

		if (_supabase == null)
		{
			SupabaseOptions options = new();
			_supabase = new Client("https://<project-id>.supabase.co",
				"<project-anon-key>",
				options);
		}

		try
		{
			_messageLabel.Text = "Signing up...";
			Session signup = await _supabase.Auth.SignUp(_lineEdit.Text, _passwordEdit.Text);
			_messageLabel.Text = $"Signed up {signup}";
		}
		catch (Exception e)
		{
			_messageLabel.Text = e.Message;
		}
	}
}
```

**Comment by acupofjose** [comment]
This is freaking awesome. Love it!
